src/Callable.h

```diff
@@ -0,0 +1,357 @@
+            // as part of the wrapper code, and we want this lean-and-mean
+            argv[idx] = value.get()->raw_buffer();
+        }
+
+        template<typename T, int Dims>
+        HALIDE_ALWAYS_INLINE void fill_slot(size_t idx, const ::Halide::Runtime::Buffer<T, Dims> &value) {
+            argv[idx] = value.raw_buffer();
+        }
+
+        HALIDE_ALWAYS_INLINE
+        void fill_slot(size_t idx, halide_buffer_t *value) {
+            argv[idx] = value;
+        }
+
+        HALIDE_ALWAYS_INLINE
+        void fill_slot(size_t idx, JITUserContext *value) {
+            auto *dest = &argv_scalar_store[idx];
+            *dest = (uintptr_t)value;
+            argv[idx] = dest;
+        }
+
+        template<typename T>
+        HALIDE_ALWAYS_INLINE void fill_slot(size_t idx, const T &value) {
+            auto *dest = &argv_scalar_store[idx];
+            *(T *)dest = value;
+            argv[idx] = dest;
+        }
+
+        template<typename T>
+        HALIDE_ALWAYS_INLINE void fill_slots(size_t idx, const T &value) {
+            fill_slot(idx, value);
+        }
+
+        template<typename First, typename Second, typename... Rest>
+        HALIDE_ALWAYS_INLINE void fill_slots(int idx, First &&first, Second &&second, Rest &&...rest) {
+            fill_slots<First>(idx, std::forward<First>(first));
+            fill_slots<Second, Rest...>(idx + 1, std::forward<Second>(second), std::forward<Rest>(rest)...);
+        }
+    };
+
+    Callable();
+    Callable(const std::string &name,
+             const JITHandlers &jit_handlers,
+             const std::map<std::string, JITExtern> &jit_externs,
+             Internal::JITCache &&jit_cache);
+
+    // Note that the first entry in argv must always be a JITUserContext*.
+    int call_argv_checked(size_t argc, const void *const *argv, const QuickCallCheckInfo *actual_cci) const;
+    int call_argv_fast(size_t argc, const void *const *argv) const;
+
+    void do_check_fail(int bad_idx, size_t argc, const char *verb) const;
+    void check_qcci(size_t argc, const QuickCallCheckInfo *actual_cci) const;
+    void check_fcci(size_t argc, const FullCallCheckInfo *actual_cci) const;
+
+    template<typename... Args>
+    int call(JITUserContext *context, Args &&...args) const {
+        // This is built at compile time!
+        static constexpr auto actual_arg_types = make_qcci_array<JITUserContext *, Args...>();
+
+        constexpr size_t count = sizeof...(args) + 1;
+        ArgvStorage<count> argv(context, std::forward<Args>(args)...);
+        return call_argv_checked(count, &argv.argv[0], actual_arg_types.data());
+    }
+
+    /** Return the expected Arguments for this Callable, in the order they must be specified, including all outputs.
+     * Note that the first entry will *always* specify a JITUserContext. */
+    const std::vector<Argument> &arguments() const;
+
+public:
+    template<typename... Args>
+    HALIDE_MUST_USE_RESULT int
```

Comment:
If the user has installed an error handler that throws an exception or aborts, must they really use the result here?

Reply:
Unless you've overriding error handling, the only value a halide pipeline can possibly return is zero. Returning non-zero values is non-standard behavior. Our AOT apps (e.g. local_laplacian) don't check the return code for this reason.

Reply:
Making it easy to ignore possible error results is... not a good idea. 

When we added the `HALIDE_MUST_USE_RESULT` to AOT results inside Google, it revealed a *lot* of places where errors were being ignored (and most of those places were using a runtime that replaced the error handler).

We are introducing an effectively entirely new API here; it would be a serious design mistake to have the default recommended style be "ignore potential errors".

Reply:
Updated per chat discussion, PTAL